syntax = "proto3";

message _TestRequest {
    uint32 id = 1;
    oneof message {
        _TestCreateIO createIO = 2;
        _TestSetIOValue setIOValue = 3;
        _TestGetIOValue getIOValue = 4;
        _TestClearIOS clearIOs = 5;
        _TestFreeMemory freeMemory = 6;
        _TestSetClockOffset setClockOffset = 7;
        _TestGetMillis getMillis = 8;
        _TestSetIOSource setIOSource = 9;
        _TestLoadAPIFromEEPROM loadAPIFromEEPROM = 10;
        _TestResetClock resetClock = 11;
    }
}

message _TestResponseValue {
    oneof value {
        bool boolValue = 2;
        int32 intValue = 3;
        uint32 uintValue = 4;
        float doubleValue = 5;
        string stringValue = 6;
    }
}

message _TestResponse {
    uint64 id = 1;
    _TestResponseValue message = 2;
    bool error = 3;
}

message _TestCreateIO {
    uint32 pin = 1;
    enum IOType {
        DIGITAL = 0;
        ANALOGIC = 1;
    }
    IOType type = 2;
}

message _TestSetIOValue {
    uint32 pin = 1;
    uint32 value = 2;
}

message _TestGetIOValue {
    uint32 pin = 1;
}

message _TestClearIOS {
}

message _TestFreeMemory {
}

message _TestSetClockOffset {
    uint32 value = 1;
}

message _TestGetMillis {
}

message _TestSetIOSource {
    enum IOSource {
        VIRTUAL = 0;
        PHYSICAL = 1;
    }
    IOSource source = 1;
}

message _TestLoadAPIFromEEPROM {

}

message _TestResetClock {
}
